### PR TITLE
Add priority-limited locks

### PR DESCRIPTION
**Pull request by bugadani**
However, when it does, we can use it in embassy-executor to lock certain resources without affecting higher priority levels. This will come in handy once we manage our own timer queue, which is only ever accessed from the executor's priority level, and the alarm's level (which we'll be able to reduce to executor+1). Priority-limiting access to the queue will result in less jitter in higher priority executors. Hopefully :)

skip-changelog is purposeful. `sync` is a hidden module, which needs to be public because embassy/wifi build on it, but this new addition is pretty experimental - API, documentation or functionality may still change I think.

### CODE REVIEW

**Comment by Dominaezzz** [comment]
I haven't taken the time to see if it matters for this yet but just in case you figure it out before I do:

Given priorities are being treated more granular-ly than just a binary on/off thing now, it might a good time to improve the priority restoration instruction here. https://github.com/esp-rs/esp-hal/blob/033c4b84c66208b38664bd21b3fcf47377406844/esp-hal-embassy/src/executor/thread.rs#L141

That should be restoring back to the original level rather than just zero.

**Comment by bugadani** [comment]
Oops I accidentally reverted that change...

**Comment by bugadani** [comment]
As someone mentioned, there is a chance for these to cause deadlocks. Back to draft for some more reasoning :)

File: esp-hal/src/sync.rs [hunk 1]
```diff
@@ -1,52 +1,150 @@
...
+                        Priority::None => core::arch::asm!("rsil {0}, 0", out(reg) token),
+                        Priority::Priority1 => core::arch::asm!("rsil {0}, 1", out(reg) token),
+                        Priority::Priority2 => core::arch::asm!("rsil {0}, 2", out(reg) token),
+                        Priority::Priority3 => core::arch::asm!("rsil {0}, 3", out(reg) token),
+                    };
+
+                    let prev_interrupt_priority = token as u8 & 0x0F;
+                } else {
+                    compile_error!("Unsupported architecture")
+                }
```
Comment (bjoernQ) [inline esp-hal/src/sync.rs:70]: I guess ideally this functionality should live in the interrupt module

**Comment by bugadani** [comment]
I see a similar enough crash on main, I don't think this is the PR's fault:

```
Failed to connect to wifi: Disconnected
About to connect...
Wifi connected!


====================== PANIC ======================
panicked at C:\Users\bugad\.cargo\registry\src\index.crates.io-6f17d22bba15001f\smoltcp-0.11.0\src\socket\dns.rs:170:49:
called `Result::unwrap()` on an `Err` value: ()
```

**Review by bjoernQ — APPROVED** [review]
LGTM

**Review by Dominaezzz — APPROVED** [review]
Xtensa side LGTM. (I don't really know the RISCV aspects of this)

File: esp-hal/src/sync.rs [hunk 2]
```diff
@@ -1,52 +1,109 @@
...
-        };
+    use crate::interrupt::Priority;
 
-        // Ensure no subsequent memory accesses are reordered to before interrupts are
-        // disabled.
-        compiler_fence(Ordering::SeqCst);
+    /// Trait for single-core locks.
+    pub trait RawLock {
+        unsafe fn enter(&self) -> critical_section::RawRestoreState;
+        unsafe fn exit(&self, token: critical_section::RawRestoreState);
```
[the remark this answers is not in the document]
Comment (bugadani) [reply esp-hal/src/sync.rs:18]: I have no clue what you mean by this, or why this comment is placed on an unrelated implementation detail.

[the remark this answers is not in the document]
Comment (Dominaezzz) [reply esp-hal/src/sync.rs:18]: Yes

**Pull request merged**